styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java

```diff
@@ -108,30 +118,29 @@ public void transitionInto(RunState state) {
     }
   }
 
-  private String dockerRunnerStart(RunState state) throws IOException {
-    final WorkflowInstance workflowInstance = state.workflowInstance();
+  private RunSpec createRunSpec(RunState state) throws ResourceNotFoundException {
     final Optional<ExecutionDescription> executionDescriptionOpt = state.data().executionDescription();
 
     final ExecutionDescription executionDescription = executionDescriptionOpt.orElseThrow(
         () -> new ResourceNotFoundException("Missing execution description for "
                                             + state.workflowInstance().toKey()));
 
+    final String executionId = state.data().executionId()
+        // Backwards compatibility fallback
+        .orElseGet(ExecutionDescriptionHandler::createExecutionId);
```

Comment:
As for pure replays, they just go through the transitions defined by `RunState`. Old WFIs will have `executionId`  added to their state by a `submitted` event, new WFIs will have it added by a `submit` event.